File: backend/ria_ml/AMDnLockboxes_process.m
function AMDnLockboxes = AMDnLockboxes_process(AMDnLockboxes, market, client);

  % get number of years of returns 
     [nscen nyrs] = size(market.cumRmsM);
  % get n
     n = AMDnLockboxes.cumRmDistributionYear;
     if n < 2 ; n = 2; end;
     if n > nyrs; n = nyrs; end; 
 
  % set lockbox proportions for initial years to investment in the market
  % portfolio
     xfs = zeros(1,n-1);
     xms = ones (1,n-1); 
  % create matrix of proportions
     xs = [xfs; xms];
     
  % do regressions to compute contents of remaining lockboxes
     for yr = n:nyrs
        % sort cumulative returns
           x = sort(market.cumRmsM(:,yr),'ascend');
           y = sort(market.cumRmsM(:,n),'ascend');
        % regress y values on x values
        %     y = b(1) + b(2)*x
           xvals = [ ones(length(x), 1) x ];
           b = xvals \ y;
        % compute lockbox contents
           xf =  b(1)/mean(market.cumRfsM(:,yr));
           xm =  b(2);
        % add to xs matrix
           xs = [xs [xf;xm]];
     end % for yr = n:nyrs  
     
  % add lockbox holdings to AMDnLockboxes
     AMDnLockboxes.proportions = xs;
  
  % plot contents if requested
     if lower(AMDnLockboxes.showProportions) == 'y'
        fig = figure;
        x = 1:1:size(xs,2);
        bar(x,xs','stacked'); grid;
        set(gca,'FontSize',30);
        ss = client.figurePosition;
        set(gcf,'Position',ss);
        set(gcf,'Color',[1 1 1]);
        xlabel('Lockbox Maturity Year ','fontsize',30);
        ylabel('Amount Invested at Inception   ','fontsize',30);
        legend('TIPS ','Market ');
        ax = axis; ax(1) = 0; ax(2) = nyrs+1; ax(3) = 0; ax(4) = 1; axis(ax);
        t = ['Lockbox Contents for Approximating Market Distribution in year ' num2str(n)];
        title(t,'Fontsize',40,'Color','b');
        beep; pause;
     end; %if lower(AMDnLockboxes.showContents) = 'y'    
     
end % function 

File: backend/ria_ml/AMDnLockboxes_process.py
import numpy as np
import matplotlib.pyplot as plt

def AMDnLockboxes_process(AMDnLockboxes, market, client):
    nscen, nyrs = market['cumRmsM'].shape
    n = AMDnLockboxes['cumRmDistributionYear']
    n = max(n, 2)
    n = min(n, nyrs)

    xfs = np.zeros(n - 1)
    xms = np.ones(n - 1)
    xs = np.vstack([xfs, xms])

    for yr in range(n, nyrs + 1):
        x = np.sort(market['cumRmsM'][:, yr - 1])
        y = np.sort(market['cumRmsM'][:, n - 1])
        xvals = np.column_stack([np.ones(len(x)), x])
        b, _, _, _ = np.linalg.lstsq(xvals, y, rcond=None)
        xf = b[0] / np.mean(market['cumRfsM'][:, yr - 1])
        xm = b[1]
        xs = np.hstack([xs, [[xf], [xm]]])

    AMDnLockboxes['proportions'] = xs

    if AMDnLockboxes['showProportions'].lower() == 'y':
        fig, ax = plt.subplots()
        x = np.arange(1, xs.shape[1] + 1)
        ax.bar(x, xs[0, :], label='TIPS')
        ax.bar(x, xs[1, :], bottom=xs[0, :], label='Market')
        ax.grid(True)
        ax.set_xlabel('Lockbox Maturity Year')
        ax.set_ylabel('Amount Invested at Inception')
        ax.legend(['TIPS', 'Market'])
        ax.set_xlim([0, nyrs + 1])
        title = f'Lockbox Contents for Approximating Market Distribution in year {n}'
        ax.set_title(title, color='b')
        plt.show()

    return AMDnLockboxes
